cpp/src/arrow/ipc/reader.cc

```diff
@@ -1088,10 +1163,33 @@ class RecordBatchFileReaderImpl : public RecordBatchFileReader {
     return Status::OK();
   }
 
+  Future<std::shared_ptr<RecordBatch>> ReadRecordBatchAsync(int i) {
+    DCHECK_GE(i, 0);
+    DCHECK_LT(i, num_record_batches());
+
+    auto cached_metadata = cached_metadata_.find(i);
+    if (cached_metadata != cached_metadata_.end()) {
+      return ReadCachedRecordBatch(i, cached_metadata->second);
+    }
+
+    return Status::Invalid(
+        "Asynchronous record batch reading is only supported after a call to "
+        "PreBufferMetadata or PreBufferBatches");
+  }
+
   Result<std::shared_ptr<RecordBatch>> ReadRecordBatch(int i) override {
     DCHECK_GE(i, 0);
     DCHECK_LT(i, num_record_batches());
 
+    auto cached_metadata = cached_metadata_.find(i);
+    if (cached_metadata != cached_metadata_.end()) {
+      return ReadCachedRecordBatch(i, cached_metadata->second).result();
+    }
+
+    // FIXME: What if they have prebuffered metadata and so the dictionary read has
+    // started but this batch wasn't prebuffered and so the dictionaries haven't been
+    // finished getting read yet.
```

Comment:
Is this FIXME a worry?

Reply:
I suppose.  Probably more me being paranoid than anything.  I added a test that exposed the failure and made a more robust "wait for dictionaries to finish reading".

I'll point out that both the old mechanism and this mechanism are not thread safe (I haven't given too much thought to multiple threads using the same reader but I think it would be a bad idea).

Reply:
That's fair. It would be weird to mix the async and sync versions too.